### PR TITLE
New cards — Congratulations, Get Well

### LINKED ISSUE #8609: New Card and Badge: Get Well
We've got a new card and its associated badge ready to implement. :)

The image files and the information for the Market item and badge name can be found on this Trello card: they are not posted here directly in order to preserve the correctly-formatted filenames. :)

https://trello.com/c/W1lhmwEp/7-get-well-card-and-badge

### LINKED ISSUE #8610: New Card and Badge: Congratulations!
We've got a new card and its associated badge ready to implement. :)

The image files and the information for the Market item and badge name can be found on this Trello card: they are not posted here directly in order to preserve the correctly-formatted filenames. :)

https://trello.com/c/ahqm9IFN/6-congratulations-card-and-badge

### PR DESCRIPTION
**Pull request by librarianmage**
[//]: # (Note: See http://habitica.wikia.com/wiki/Using_Habitica_Git#Pull_Request for more info)

[//]: # (Put Issue # or URL here, if applicable. This will automatically close the issue if your PR is merged in)
Fixes #8610 
Fixes #8609 

### Changes
[//]: # (Describe the changes that were made in detail here. Include pictures if necessary)

[//]: # (Put User ID in here - found in Settings -> API)

----
UUID:

### CODE REVIEW

**Comment by librarianmage** [comment]
:( How can I fix the decrease in coverage?

**Comment by TheHollidayInn** [comment]
Hmm. I think it may be checking coverage on the content files. I don't think you should worry.

File: website/common/locales/en/generic.json [hunk 1]
```diff
@@ -179,6 +179,16 @@
...
+    "congratsCard": "Congratulations Card",
+    "congratsCardExplanation": "You both recieve the Congratulatory Companion achievement!",
+    "congratsCardNotes": "Send a Congratulations card to a party member.",
+    "congrats0": "Congratulations on your success!",
+    "congrats1": "I'm so proud of you!",
+    "congrats2": "Well done!",
+    "congrats3": "A round of applause for you!",
+    "congrats4": "Bask in your well-deserved success!",
+    "congratsCardAchievementTitle": "Congratulatory Companion",
+    "congratsCardAchievementText": "It's great to celebrate your friends' achievements! Sent or received <%= count %> Congratulations Cards.",
```
Comment (khipkin) [inline website/common/locales/en/generic.json]: nit: lowercase "congratulations cards" for consistency with other cards.

**Comment by librarianmage** [comment]
@khipkin Updated.

**Comment by librarianmage** [comment]
This PR is no longer a work in progress (well, it probably still is, but I have added everything).

File: website/common/script/content/spells.js [hunk 2]
```diff
@@ -514,6 +514,60 @@ spells.special = {
...
+        each([user, target], (u) => {
+          if (!u.achievements.congrats) u.achievements.congrats = 0;
+          u.achievements.congrats++;
+        });
+      }
+
+      if (!target.items.special.congratsReceived) target.items.special.congratsReceived = [];
+      target.items.special.congratsReceived.push(user.profile.name);
+
+      if (!target.flags) target.flags = {};
```
Comment (paglias) [inline website/common/script/content/spells.js:539]: target.flags should always exist

File: website/common/script/content/spells.js [hunk 3]
```diff
@@ -514,6 +514,60 @@ spells.special = {
       user.stats.gp -= 10;
     },
   },
+  congrats: {
```
Comment (paglias) [inline website/common/script/content/spells.js:517]: what about a test for this?

File: website/common/script/content/spells.js [hunk 4]
```diff
@@ -514,6 +514,60 @@ spells.special = {
...
+      if (!target.items.special.congratsReceived) target.items.special.congratsReceived = [];
+      target.items.special.congratsReceived.push(user.profile.name);
+
+      if (!target.flags) target.flags = {};
+      target.flags.cardReceived = true;
+
+      user.stats.gp -= 10;
+    },
+  },
+  getwell: {
```
Comment (paglias) [inline website/common/script/content/spells.js:545]: test for this?

File: website/common/script/content/spells.js [hunk 5]
```diff
@@ -514,6 +514,60 @@ spells.special = {
...
+        each([user, target], (u) => {
+          if (!u.achievements.getwell) u.achievements.getwell = 0;
+          u.achievements.getwell++;
+        });
+      }
+
+      if (!target.items.special.getwellReceived) target.items.special.getwellReceived = [];
+      target.items.special.getwellReceived.push(user.profile.name);
+
+      if (!target.flags) target.flags = {};
```
Comment (paglias) [inline website/common/script/content/spells.js:567]: flags should always exist

File: website/common/script/content/spells.js [hunk 6]
```diff
@@ -514,6 +514,60 @@ spells.special = {
...
+    immediateUse: true,
+    silent: true,
+    target: 'user',
+    notes: t('getwellCardNotes'),
+    cast (user, target) {
+      if (user === target) {
+        user.achievements.getwell++;
+      } else {
+        each([user, target], (u) => {
+          if (!u.achievements.getwell) u.achievements.getwell = 0;
```
Comment (paglias) [inline website/common/script/content/spells.js]: `achievements.getwell` shouldn't be added in the user schema?

File: website/common/script/content/spells.js [hunk 7]
```diff
@@ -514,6 +514,60 @@ spells.special = {
...
+    immediateUse: true,
+    silent: true,
+    target: 'user',
+    notes: t('congratsCardNotes'),
+    cast (user, target) {
+      if (user === target) {
+        user.achievements.congrats++;
+      } else {
+        each([user, target], (u) => {
+          if (!u.achievements.congrats) u.achievements.congrats = 0;
```
Comment (paglias) [inline website/common/script/content/spells.js]: achievements.getwell shouldn't be added in the user schema?

File: website/common/script/content/spells.js [hunk 8]
```diff
@@ -514,6 +514,60 @@ spells.special = {
...
+        user.achievements.getwell++;
+      } else {
+        each([user, target], (u) => {
+          if (!u.achievements.getwell) u.achievements.getwell = 0;
+          u.achievements.getwell++;
+        });
+      }
+
+      if (!target.items.special.getwellReceived) target.items.special.getwellReceived = [];
+      target.items.special.getwellReceived.push(user.profile.name);
```
Comment (paglias) [inline website/common/script/content/spells.js:565]: getWellReceived shouldn't be added to the user schema?

File: website/common/script/content/spells.js [hunk 9]
```diff
@@ -514,6 +514,60 @@ spells.special = {
...
+        user.achievements.congrats++;
+      } else {
+        each([user, target], (u) => {
+          if (!u.achievements.congrats) u.achievements.congrats = 0;
+          u.achievements.congrats++;
+        });
+      }
+
+      if (!target.items.special.congratsReceived) target.items.special.congratsReceived = [];
+      target.items.special.congratsReceived.push(user.profile.name);
```
Comment (paglias) [inline website/common/script/content/spells.js:537]: congratsReceived shouldn't be added to the user schema?

**Review by paglias — CHANGES_REQUESTED** [review]

**Comment by librarianmage** [comment]
I'm sorry, @paglias, but I don't understand some of your comments. I copied and pasted the code from the greeting card, so the achievement stuff is copied from that. I also don't see any tests for the other cards.

**Comment by paglias** [comment]
@MathWhiz no worries, regarding tests it's true that there aren't but I think we should start to write them for a new spells. And regarding the achiements I'm saying that you're setting values to fields that are not defined in /website/server/models/user/schema.js so they won't be saved in the database

**Comment by librarianmage** [comment]
What do you mean about the flags?

**Comment by paglias** [comment]
Same thing, they  have to be defined in the schema or they wont be saved

Il 09 Apr 2017 20:23, "MathWhiz" <notifications@github.com> ha scritto:

**Comment by khipkin** [comment]
Looks like only half of the new images have transparent backgrounds.

**Comment by librarianmage** [comment]
Whelp, I will fix that as soon as I can

**Comment by librarianmage** [comment]
@paglias 

In regards to tests, here is what I have planned:

1. Cards should return information specified in the [APIdoc](https://habitica.com/apidoc/#api-User-UserBuySpecialSpell)
2. Cards should only work on a user (not sure if done client-side)
3. Cards should increase the achievement for both users when a card is sent
4. Cards should increase the achievement once if a user sends it to themself
5. Returns error with lack of gold
6. I don't know if this is possible, but making sure that spells are only castable when "enabled"

Is there anything I am missing?

**Comment by paglias** [comment]
@MathWhiz looks like a complete list!

**Comment by librarianmage** [comment]
@paglias If you have some time, some help would be greatly appreciated. Tests have never been my forte, and I seem to have figured out how to screw them up.

File: test/api/v3/integration/user/POST-user_class_cast_spellId.test.js [hunk 10]
```diff
@@ -221,6 +221,20 @@ describe('POST /user/class/cast/:spellId', () => {
     expect(syncedGroupTask.value).to.equal(0);
   });
 
+  it('increases both user\'s achievement values', async () => {
+    let recipient = await generateUser();
+    await user.update({'stats.gp': 10});
+    await user.post(`/user/cast/cast/birthday?targetId=${recipient._id}`);
```
Comment (paglias) [inline test/api/v3/integration/user/POST-user_class_cast_spellId.test.js]: double /cast in the url

File: test/api/v3/integration/user/POST-user_class_cast_spellId.test.js [hunk 11]
```diff
@@ -221,6 +221,20 @@ describe('POST /user/class/cast/:spellId', () => {
...
+    expect(user.achievements.birthday).to.equal(1);
+    expect(recipient.achievements.birthday).to.equal(1);
+  });
+
+  it('only increases user\'s achievement one if target == caster', async () => {
+    await user.update({'stats.gp': 10});
+    await user.post(`/user/cast/cast/birthday?targetId=${user._id}`);
```
Comment (paglias) [inline test/api/v3/integration/user/POST-user_class_cast_spellId.test.js]: double /cast in the url

**Review by paglias — CHANGES_REQUESTED** [review]

**Comment by librarianmage** [comment]
Finished! @paglias

**Comment by librarianmage** [comment]
It would be greatly appreciated if this pull could be reviewed. Thanks!

**Review by paglias — APPROVED** [review]

**Comment by paglias** [comment]
Looks good to me! Sorry for the delay @MathWhiz .

@SabreCat can we merge this or any other work is needed?

**Comment by paglias** [comment]
Note: I haven't tested it locally, only reviewed the code

**Comment by SabreCat** [comment]
I'll look it over, thanks!

**Comment by librarianmage** [comment]
Thank you for looking it over, @SabreCat!

**Comment by librarianmage** [comment]
@SabreCat I have made some logic fixes, so you will want to pull again to make sure everything is a-ok.

**Comment by SabreCat** [comment]
Couple of things here:

The card images in the shop aren't on the right canvas size, so they have inconsistent button sizes vs. the other ones--
image
Looks like they need to be recanvased to 57x54.

Secondly, there are enough cards after these additions that I don't think these belong in the "Special" category of the Market anymore. We should move all the cards out to their own "Cards" Market category, above Special.

**Review by SabreCat — CHANGES_REQUESTED** [review]
Recanvas, create dedicated Market category (though we can definitely help with the latter if the path forward there is unclear!)

**Comment by librarianmage** [comment]
Ok! I can take a look at resizing them/moving them

**Comment by SabreCat** [comment]
One other thing I just noticed--it's best not to include changes to any of the compiled spritesheet data (PNG and CSS) in the PR. Those changes will inevitably get overwritten in production, and it makes the PR harder to review. So if those changes can be reverted to `develop` state on this PR, it'd be a help!

**Comment by librarianmage** [comment]
Got it! I will revert them in my next PR

**Comment by librarianmage** [comment]
@SabreCat I tried resetting it, but it didn't seem to work properly.

**Comment by SabreCat** [comment]
Try

```
git checkout develop website/client/assets/* website/static/*
git commit
git push
```

**Comment by librarianmage** [comment]
I did :/

**Comment by SabreCat** [comment]
Make sure you've checked out and pulled `develop` from the main repo first, perhaps?

**Comment by SabreCat** [comment]
If you can't get that to work, I just pushed https://github.com/HabitRPG/habitica/tree/MathWhiz/new-cards -- you can cherry-pick f004256.

**Comment by librarianmage** [comment]
I'm sorry, but to make sure I understand, can you be a bit more specific
about how to do that?

I have also created the new section:

**Comment by SabreCat** [comment]
It depends a bit how you have your remotes set up locally. It'd look something like the following, supposing `HabitRPG` is the official repo:

```
git fetch HabitRPG
git checkout develop
git pull HabitRPG develop
git checkout new-cards
git checkout develop website/client/assets/* website/static/*
git commit
git push
```

**Comment by librarianmage** [comment]
Won't that also pull stuff I don't want changed?

**Comment by SabreCat** [comment]
Or if you mean the `cherry-pick` strategy:

```
git fetch HabitRPG
git checkout new-cards
git cherry-pick f004256
git push
```

**Comment by librarianmage** [comment]
@SabreCat Looks like It worked!

**Comment by SabreCat** [comment]
Success! 14 files changed, so much easier to work with than 50. Thank you!

**Comment by librarianmage** [comment]
You are welcome!

**Comment by SabreCat** [comment]
Getting this error when attempting to send cards to party members, even existing cards:

image

Are you able to complete the workflow locally?

**Comment by librarianmage** [comment]
It worked for me :/

Oh dear

File: website/common/script/content/spells.js [hunk 12]
```diff
@@ -452,7 +450,6 @@ spells.special = {
       if (!target.items.special.greetingReceived) target.items.special.greetingReceived = [];
       target.items.special.greetingReceived.push(user.profile.name);
 
-      if (!target.flags) target.flags = {};
```
Comment (SabreCat) [inline website/common/script/content/spells.js]: I think this is the source of the error I'm hitting. If the target doesn't have the `flags` object, we hit an error when trying to set a property on it. Why were these lines removed?

**Comment by librarianmage** [comment]
@paglias said that they would always exist, there fore to remove them (if I
understood them correctly)

**Comment by SabreCat** [comment]
Heh, no, they were saying that if you're saving a value to the database, it needs to be defined in the user schema, or it won't be saved. For achievements, special items, and flags. If all of those are there, you're cool.

**Comment by librarianmage** [comment]
> target.flags should always exist
?

**Comment by SabreCat** [comment]
I don't think it's good practice to rely on that, no, especially given I've encountered an error with it in my own testing.

**Comment by librarianmage** [comment]
I was just quoting what @paglias commented.

**Comment by SabreCat** [comment]
I can't even find where he said that, sorry :\

**Comment by librarianmage** [comment]
It was the fourth review he made in the large batch of them (collapsed)

**Comment by SabreCat** [comment]
Ah, I see it. In any case, perhaps he was mistaken; in common code, it's possible that it's present on the server but not the client, etc.

**Comment by librarianmage** [comment]
@SabreCat Ok, I will change it back

**Comment by librarianmage** [comment]
@SabreCat ... and done!

**Comment by SabreCat** [comment]
OK, looks great now! I have one thing I'll probably tweak myself, moving these achievements out of Seasonal to Basic, but it's a preexisting issue with the Greeting and Thank You cards, not necessary for this PR. I'll tag this for merge and credit it to your Blacksmith progress!

**Comment by librarianmage** [comment]
:tada: Thanks!

**Comment by librarianmage** [comment]
Slowly making progress towards my next tier!

**Review by SabreCat — APPROVED** [review]

**Pull request merged**